src/runtime_src/core/pcie/tools/xbmgmt/xspi.cpp

```diff
@@ -1604,6 +1689,193 @@ int XSPI_Flasher::revertToMFG(void)
         std::cout << "ERROR: Unable to set bitstream guard!" << std::endl;
         return -EINVAL;
     }
+    return 0;
+}
+
+static int splitMcsLine(std::string& line,
+    unsigned int& type, unsigned int& address, std::vector<unsigned char>& data)
+{
+    int ret = 0;
+
+    // Every line should start with ":"
+    if (line[0] != ':')
+        return -EINVAL;
+
+    unsigned int len = std::stoi(line.substr(1, 2), NULL , 16);
+    type = std::stoi(line.substr(7, 2), NULL , 16);
+    address = std::stoi(line.substr(3, 4), NULL, 16);
+    data.clear();
+
+    switch (type) {
+    case 0:
+        if (len > 16) {
+            // For xilinx mcs files data length should be 16 for all records
+            // except for the last one which can be smaller
+            ret = -EINVAL;
+        } else {
+            unsigned int l = len * 2;
+            std::string d = line.substr(9, len * 2);
+            for (unsigned int i = 0; i < l; i += 2)
+                data.push_back(std::stoi(d.substr(i, 2), NULL, 16));
+        }
+        break;
+    case 1:
+        break;
+    case 4:
+        if (len != 2) {
+            // For xilinx mcs files extended address can only be 2 bytes
+            ret = -EINVAL;
+        }
+        address = std::stoi(line.substr(9, len * 2), NULL, 16);
+        break;
+    default:
+        // Xilinx mcs files should not contain other types
+        ret = -EINVAL;
+        break;
+    }
+
+    return ret;
+}
+
+static inline unsigned int pageOffset(unsigned int addr)
+{
+    return (addr & 0xffff);
+}
+
+static int mcsStreamToBin(std::istream& mcsStream, unsigned int& currentAddr,
+    std::vector<unsigned char>& buf, unsigned int& nextAddr)
+{
+    bool done = false;
+    size_t cnt = 0;
+
+    buf.clear();
+
+    while (!mcsStream.eof() && !done) {
+        std::string line;
+        unsigned int type;
+        unsigned int addr;
+        std::vector<unsigned char> data;
+
+        std::getline(mcsStream, line);
+        if (line.size() == 0)
+            continue;
+
+        if (splitMcsLine(line, type, addr, data) != 0) {
+            std::cout << "Found invalid MCS line: " << line << std::endl;
+            return -EINVAL;
+        }
+
+        switch (type) {
+        case 0:
+            if (currentAddr == UINT_MAX) {
+                std::cout << "MCS missing page starting address" << std::endl;
+                return -EINVAL;
+            } else if (buf.size() == 0) {
+                // we're the 1st data in this page, updating the current addr
+                assert(pageOffset(currentAddr) == 0);
+                currentAddr |= addr;
+            } else if (pageOffset(currentAddr + buf.size()) != addr) {
```

Comment:
Will this fail if the MCS contains multiple images? For example the golden image MCS file contains multiple bitstreams and they are not contiguous and will have a gap between the two bitstreams.

Reply:
The main point I'm making here is an MCS does not map directly to a single bitstream. The MCS file can contain multiple images at different offsets.

Reply:
it should work, but i have not tested since i don't have a mcs with more than one chunk to test with. but if there is a gap, you'd expect to see a type 4 line, right? type 0 lines should be contiguous, no?

Reply:
Ah I missed your logic checking in the case 4 statement below. Looks like it would work then. You can test this with the golden image MCS files as those have 2 bitstreams inside that aren't contiguous. You'd have to make sure to unlock the flash first so that you can erase the golden image.